syntax = "proto3";

package operationTransportation;

service clientService {
    // send file from server to client
    rpc initialize(empty) returns (file_from_server);
    // send operation from client to server
    rpc sendOP(editor_request) returns (last_executed_operation);
    // send command to write from vector to file
    rpc writeToFile(empty) returns (last_executed_operation);
    // request for last operation
}

// used operations
enum OP_type {
    INSERT = 0;
    DELETE = 1;
    UNDO = 2;
    REDO = 3;
    ADD_LINE = 4;
    DEL_LINE = 5;
    INVALID = 6;
}

// message with information for operation
message editor_request {
    OP_type op = 1;
    uint64 pos = 2;
    uint64 line = 3;
    uint32 sym = 4;
    uint64 user_id = 5;
    uint64 op_id = 6;
}

message last_executed_operation {
    uint64 op_id = 1;
}

// empty message for void and no-argument calls
message empty {}

// vector of strings which represent the file
message file_from_server {
    repeated string file = 1;
    uint64 op_id = 2;
}
